feat: retrieve deployed contract

lgalabru opened the pull request.

MicaiahReid commented on addons/evm/src/commands/actions/call_contract.rs:
```diff
@@ -48,91 +48,106 @@ lazy_static! {
                 typing: Type::addon(EVM_ADDRESS),
                 optional: true,
-                interpolable: true
+                tainting: true,
```
I'm not sure if this should be tainting - it is just used for type checking, to improve the error specificity. Chances are, if you change this, you'll be changing some _other_ field that _is_ tainting, but this one itself is not.

MicaiahReid replied:
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/evm/src/commands/actions/call_contract.rs:
```diff
@@ -48,91 +48,106 @@ lazy_static! {
                 typing: Type::string(),
                 optional: true,
-                interpolable: true
+                tainting: true,
```
I'm not sure the type should be tainting. This is more _how_ the tx call or deployment is made (specifically centered around gas pricing of the call) - if it's already been invoked/created with a given type I don't think the user cares about the type any more

MicaiahReid replied:
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/evm/src/commands/actions/call_contract.rs:
```diff
@@ -48,91 +48,106 @@ lazy_static! {
                 typing: Type::integer(),
                 optional: true,
-                interpolable: true
+                tainting: true,
```
Don't think this is tainting

MicaiahReid replied:
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/evm/src/commands/actions/call_contract.rs:
```diff
@@ -48,91 +48,106 @@ lazy_static! {
                 typing: Type::integer(),
                 optional: true,
-                interpolable: true
+                tainting: true,
```
Don't think this is tainting

MicaiahReid replied:
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/evm/src/commands/actions/check_confirmations.rs:
```diff
@@ -25,28 +25,32 @@ lazy_static! {
+                    documentation: "The URL of the EVM API used to poll for the transaction's inclusion in a block.",
+                    typing: Type::string(),
+                    optional: false,
+                    tainting: true,
```
It's interesting that the api url isn't tainting for call_contract, but it is for this - is that on purpose?

lgalabru replied:
Addressed with latest

MicaiahReid commented on addons/evm/src/commands/actions/deploy_contract.rs:
```diff
@@ -47,97 +47,113 @@ lazy_static! {
                 typing: Type::string(),
                 optional: true,
-                interpolable: true
+                tainting: true,
```
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/evm/src/commands/actions/deploy_contract.rs:
```diff
@@ -47,97 +47,113 @@ lazy_static! {
                 typing: Type::integer(),
                 optional: true,
-                interpolable: true
+                tainting: true,
```
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/evm/src/commands/actions/deploy_contract.rs:
```diff
@@ -47,97 +47,113 @@ lazy_static! {
                 typing: Type::integer(),
                 optional: true,
-                interpolable: true
+                tainting: true,
```
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/evm/src/commands/actions/deploy_contract.rs:
```diff
@@ -47,97 +47,113 @@ lazy_static! {
                 typing: Type::integer(),
                 optional: true,
-                interpolable: true
+                tainting: true,
```
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/evm/src/commands/actions/deploy_contract.rs:
```diff
@@ -47,97 +47,113 @@ lazy_static! {
                 typing: Type::integer(),
                 optional: true,
-                interpolable: true
+                tainting: true,
```
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/evm/src/commands/actions/deploy_contract.rs:
```diff
@@ -47,97 +47,113 @@ lazy_static! {
                 typing: Type::integer(),
                 optional: true,
-                interpolable: true
+                tainting: false,
```
Confirmations is tainting on the check_confirmations function, but not here. Is that right?

MicaiahReid commented on addons/evm/src/commands/actions/deploy_contract_create2.rs:
```diff
@@ -50,133 +50,155 @@ lazy_static! {
                 typing: Type::string(),
                 optional: true,
-                interpolable: true
+                tainting: true,
```
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/evm/src/commands/actions/deploy_contract_create2.rs:
```diff
@@ -50,133 +50,155 @@ lazy_static! {
                 typing: Type::integer(),
                 optional: true,
-                interpolable: true
+                tainting: true,
```
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/evm/src/commands/actions/deploy_contract_create2.rs:
```diff
@@ -50,133 +50,155 @@ lazy_static! {
                 typing: Type::integer(),
                 optional: true,
-                interpolable: true
+                tainting: true,
```
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/evm/src/commands/actions/deploy_contract_create2.rs:
```diff
@@ -50,133 +50,155 @@ lazy_static! {
                 typing: Type::integer(),
                 optional: true,
-                interpolable: true
+                tainting: true,
```
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/evm/src/commands/actions/deploy_contract_create2.rs:
```diff
@@ -50,133 +50,155 @@ lazy_static! {
                 typing: Type::integer(),
                 optional: true,
-                interpolable: true
+                tainting: true,
```
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/evm/src/commands/actions/eth_call.rs:
```diff
@@ -30,91 +30,106 @@ lazy_static! {
                 typing: Type::string(),
                 optional: true,
-                interpolable: true
+                tainting: true,
```
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/evm/src/commands/actions/eth_call.rs:
```diff
@@ -30,91 +30,106 @@ lazy_static! {
                 typing: Type::integer(),
                 optional: true,
-                interpolable: true
+                tainting: true,
```
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/evm/src/commands/actions/eth_call.rs:
```diff
@@ -30,91 +30,106 @@ lazy_static! {
                 typing: Type::integer(),
                 optional: true,
-                interpolable: true
+                tainting: true,
```
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/evm/src/commands/actions/eth_call.rs:
```diff
@@ -30,91 +30,106 @@ lazy_static! {
                 typing: Type::integer(),
                 optional: true,
-                interpolable: true
+                tainting: true,
```
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/evm/src/commands/actions/eth_call.rs:
```diff
@@ -30,91 +30,106 @@ lazy_static! {
                 typing: Type::integer(),
                 optional: true,
-                interpolable: true
+                tainting: true,
```
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/evm/src/commands/actions/sign_transfer.rs:
```diff
@@ -42,79 +42,85 @@ lazy_static! {
                 typing: Type::string(),
                 optional: true,
-                interpolable: true
+                tainting: true,
```
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/evm/src/commands/actions/sign_transfer.rs:
```diff
@@ -42,79 +42,85 @@ lazy_static! {
                 typing: Type::integer(),
                 optional: true,
-                interpolable: true
+                tainting: true,
```
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/evm/src/commands/actions/sign_transfer.rs:
```diff
@@ -42,79 +42,85 @@ lazy_static! {
                 typing: Type::integer(),
                 optional: true,
-                interpolable: true
+                tainting: true,
```
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/evm/src/commands/actions/sign_transfer.rs:
```diff
@@ -42,79 +42,85 @@ lazy_static! {
                 typing: Type::integer(),
                 optional: true,
-                interpolable: true
+                tainting: true,
```
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/evm/src/commands/actions/sign_transfer.rs:
```diff
@@ -42,79 +42,85 @@ lazy_static! {
                 typing: Type::integer(),
                 optional: true,
-                interpolable: true
+                tainting: true,
```
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/sp1/src/commands/actions/create_proof.rs:
```diff
@@ -24,25 +24,29 @@ lazy_static! {
                     typing: Type::string(),
                     optional: true,
-                    interpolable: true
+                    tainting: true,
```
```suggestion
                    tainting: false,
```
If I've already verified successfully locally, then decide on the next one I want to use a key, I don't think setting it should rerun

MicaiahReid commented on addons/stacks/src/commands/actions/broadcast_transaction.rs:
```diff
@@ -32,38 +32,44 @@ lazy_static! {
                     typing: Type::string(),
                     optional: true,
-                    interpolable: true
+                    tainting: true,
```
I thought in others the rpc url was false?

MicaiahReid replied:
```suggestion
                    tainting: false,
```

MicaiahReid commented on crates/txtx-addon-kit/src/types/signers.rs:
```diff
@@ -292,6 +292,18 @@ impl SignerInstance {
+        for input in self.specification.inputs.iter() {
+            let Some(value) = evaluated_inputs.inputs.get_value(&input.name) else { continue };
+            if input.sensitive {
+                println!("Including {} for fingerprint", input.name);
```
```suggestion
```

MicaiahReid commented on addons/stacks/src/commands/actions/send_stx.rs:
```diff
@@ -40,55 +40,64 @@ lazy_static! {
                     typing: Type::string(),
                     optional: true,
-                    interpolable: true
+                    tainting: true,
```
```suggestion
                    tainting: false,
```

MicaiahReid commented on addons/stacks/src/commands/actions/send_stx.rs:
```diff
@@ -40,55 +40,64 @@ lazy_static! {
                     typing: Type::string(),
                     optional: true,
-                    interpolable: true
+                    tainting: true,
```
```suggestion
                    tainting: false,
```

MicaiahReid commented on addons/stacks/src/commands/actions/send_stx.rs:
```diff
@@ -40,55 +40,64 @@ lazy_static! {
                     typing: Type::integer(),
                     optional: true,
-                    interpolable: true
+                    tainting: true,
```
```suggestion
                    tainting: false,
```

MicaiahReid commented on addons/stacks/src/commands/actions/deploy_contract.rs:
```diff
@@ -148,48 +165,56 @@ lazy_static! {
                     typing: Type::string(),
                     optional: true,
-                    interpolable: true
+                    tainting: true,
```
```suggestion
                    tainting: false,
```

MicaiahReid commented on addons/stacks/src/commands/actions/deploy_contract.rs:
```diff
@@ -148,48 +165,56 @@ lazy_static! {
+                    tainting: true,
+                    internal: false
                 },
                 contracts_ids_lazy_dependencies: {
```
I think we need a better name for this field and the above `contracts_ids_dependencies`. It's weird that all three words are plural.

`dependency_contract_ids` & `lazy_dependency_contract_ids`? Is "lazy dependency" a known concept? I've never heard of that before

MicaiahReid commented on addons/stacks/src/commands/actions/deploy_contract.rs:
```diff
@@ -57,88 +60,102 @@ lazy_static! {
                     typing: Type::integer(),
                     optional: true,
-                    interpolable: true
+                    tainting: true,
```
```suggestion
                    tainting: false,
```

MicaiahReid commented on addons/stacks/src/commands/actions/sign_transaction.rs:
```diff
@@ -53,43 +53,50 @@ lazy_static! {
                 typing: Type::integer(),
                 optional: false,
-                interpolable: true
+                tainting: true,
```
```suggestion
                tainting: false,
```

MicaiahReid commented on addons/telegram/src/send_message.rs:
```diff
@@ -22,25 +22,29 @@ lazy_static! {
               typing: Type::string(),
               optional: false,
-              interpolable: true
+              tainting: true,
```
```suggestion
              tainting: false,
```

MicaiahReid commented on addons/telegram/src/setup_chat.rs:
```diff
@@ -22,13 +22,15 @@ lazy_static! {
               typing: Type::string(),
               optional: false,
-              interpolable: true
+              tainting: true,
```
```suggestion
              tainting: false,
```

MicaiahReid reviewed (changes requested):
Sorry this is a lot 😅

Also, what's the `internal` flag for?

MicaiahReid commented:
Any thoughts on making `tainting`/`internal` optional fields for the macro with a default value? It would be nice when making these, but I admit you run the risk of forgetting to set the field.

MicaiahReid commented:
One last complaint: I really don't love the word "tainting" for this. It feels like internal kitchen stuff that driver authors shouldn't need to be aware of? Would something like `critical` work?

lgalabru commented:
@MicaiahReid

> what's the internal flag for?

Internal stands for inputs needed internally, that we don't want to expose in the documentation

> dependency_contract_ids & lazy_dependency_contract_ids

Addressed with latest commit

> Is "lazy dependency" a known concept?

In this context, it means that the dependencies can be in the same block - vs committed in a prior block.

> It's interesting that the api url isn't tainting for call_contract, but it is for this

Addressed with latest commit
 
> It feels like internal kitchen stuff that driver authors shouldn't need to be aware of?

Unfortunately this work needs to be done manually, and a default value would lead to bugs. I don't see how we can infer this, but could be missing something 

>  I really don't love the word "tainting" for this

The tainting concept / terminology is coming from Terraform where tainted resources are detected automatically but can also be done manually through a command line - `txtx taint <runbook> <action>` for forcing the re-execution of a subgraph. Open to another idea, but I don't see `critical` as an adequate alternative.

MicaiahReid commented:
> Internal stands for inputs needed internally, that we don't want to expose in the documentation

Very nice 💯 

>The tainting concept / terminology is coming from Terraform where tainted resources are detected automatically but can also be done manually through a command line - `txtx taint <runbook> <action>` for forcing the re-execution of a subgraph. Open to another idea, but I don't see `critical` as an adequate alternative.

Okay, understood. How about `can_taint_execution` or `can_taint_state`? The `ing` suffix feels off, I'd love to have it worded more like a bool. After that, I'm good!

lgalabru commented:
We're currently using adjectives for qualifying inputs (`internal` vs `is_internal` , `optional` vs `is_optional`, etc). 
Can you think of an alternative adjective for qualifying these `tainting` inputs?

MicaiahReid commented:
Sure, but I'd argue that a bool like `internal`/`optional` is much clearer in this case - they say what this input _is_, while tainting describes what it _does_. 

Since as you said tainting is a specific concept with Terraform, I don't think we should look for another adjective, but rather word it in a way that is less awkward and makes the purpose of the field clearer.
